examples.md

```diff
@@ -99,6 +100,25 @@ We cache the elements of the Cabal store separately, as the entirety of `~/.caba
     key: ${{ runner.os }}-${{ matrix.ghc }}-dist-newstyle
 ```
 
+## Haskell - Stack
+
+```yaml
+- uses: actions/cache@v1
+  name: Cache ~/.stack
+  with:
+    path: ~/.stack
+    key: ${{ runner.os }}-stack-global-${{ hashFiles('stack.yaml') }}-${{ hashFiles('package.yaml') }}
+    restore-keys: |
+      ${{ runner.os }}-stack-global-
+- uses: actions/cache@v1
+  name: Cache .stack-work
+  with:
+    path: .stack-work
+    key: ${{ runner.os }}-stack-work-${{ hashFiles('stack.yaml') }}-${{ hashFiles('package.yaml') }}-${{ '**/*.hs' }}
```

Comment:
`.hs` are source files. Is it a good or bad idea to create hash over all sources?

Reply:
Actually yes, I'd disagree with that as well. [In my own post on the matter](https://www.fosskers.ca/en/blog/github-ci#orga67e3a) I recommend just considering the hash of the `stack.yaml`.

Reply:
Interesting :) What about `package.yaml`? All dependencies are listed there (which seems to be the equivalent to `package.json` or `pom.xml`). My `stack.yaml(.lock)` mainly contains the resolver. With restore-keys it'll fall back to use a cache that is defined by the `stack.yaml`. Would you advice against using `package.yaml` (with `hpack`)?

Btw. I wanted to ask for feedback in my own PR, but mixed those two up ☺️ https://github.com/actions/cache/pull/236#discussion_r400489476

Reply:
@fosskers, your post describes caching the `~/.stack` folder, and I definitely agree having hashes in the restore key related to source files doesn't make any sense for that. The example I'm adding with this PR has a cache for both the global `~/.stack` folder, and the projects `.stack-work` folder. 

For the global cache (`~/.stack`), I agree with @andys8 that both `stack.yaml` and `package.yaml` should be included in the key, since changes to either will effect that cache.

For the project cache (`.stack-work`), I'm pretty sure we want to include all source files in the key since changes to those files are the main thing that change the contents of that cache.

Reply:
Perhaps there's a discussion to be had though about whether the example should include a cache for `.stack-work`.

I definitely find it helpful to include in my workflows, especially for large projects. I'm very keen for my workflows to run as quickly as possible, and not caching `.stack-work` means the workflow will always have the recompile the project from scratch.

If we are going to cache `.stack-work` then having the key include a hash of all the source files seems like the right call, since the contents of the cache will likely change if any of those files change. If we only include hashes for the `.yaml` configuration files, then the cache will only update when those files change, which is much less often than changes to the source files. As a result, the cache will become stale pretty quickly, and workflow runtimes will increase, since more of the project will need to be recompiled on each workflow run.

Reply:
Just looked through my open PR and realized this was still pending. Any thoughts on my comments?